lib/db/goleveldb.go

```diff
@@ -0,0 +1,289 @@
+	*leveldb.DB
+	closed   bool
+	closeMut *sync.RWMutex
+	iterWG   sync.WaitGroup
+}
+
+// OpenGoleveldb attempts to open the database at the given location, and runs
+// recovery on it if opening fails. Worst case, if recovery is not possible,
+// the database is erased and created from scratch.
+func OpenGoleveldb(location string) (Backend, error) {
+	opts := &opt.Options{
+		OpenFilesCacheCapacity: goleveldbMaxOpenFiles,
+		WriteBuffer:            goleveldbWriteBuffer,
+	}
+	return openGoleveldb(location, opts)
+}
+
+// NewLowlevelFromGoleveldb opens the db and wraps it in a Lowlevel.
+func NewLowlevelFromGoleveldb(location string) (*Lowlevel, error) {
+	db, err := OpenGoleveldb(location)
+	if err != nil {
+		return nil, err
+	}
+	return NewLowlevel(db), nil
+}
+
+// NewROLowlevelFromGoleveldb attempts to open the database at the given location, read only.
+func NewROLowlevelFromGoleveldb(location string) (*Lowlevel, error) {
+	opts := &opt.Options{
+		OpenFilesCacheCapacity: goleveldbMaxOpenFiles,
+		ReadOnly:               true,
+	}
+	db, err := openGoleveldb(location, opts)
+	if err != nil {
+		return nil, err
+	}
+	return NewLowlevel(db), nil
+}
+
+func openGoleveldb(location string, opts *opt.Options) (*goleveldb, error) {
+	db, err := leveldb.OpenFile(location, opts)
+	if leveldbIsCorrupted(err) {
+		db, err = leveldb.RecoverFile(location, opts)
+	}
+	if leveldbIsCorrupted(err) {
+		// The database is corrupted, and we've tried to recover it but it
+		// didn't work. At this point there isn't much to do beyond dropping
+		// the database and reindexing...
+		l.Infoln("Database corruption detected, unable to recover. Reinitializing...")
+		if err := os.RemoveAll(location); err != nil {
+			return nil, errorSuggestion{err, "failed to delete corrupted database"}
+		}
+		db, err = leveldb.OpenFile(location, opts)
+	}
+	if err != nil {
+		return nil, errorSuggestion{err, "is another instance of Syncthing running?"}
+	}
+	return &goleveldb{
+		DB:       db,
+		closeMut: &sync.RWMutex{},
+		iterWG:   sync.WaitGroup{},
+	}, nil
+}
+
+// OpenMemory returns a new Lowlevel referencing an in-memory database.
+func OpenMemory() *Lowlevel {
+	db, _ := leveldb.Open(storage.NewMemStorage(), nil)
+	return NewLowlevel(&goleveldb{
+		DB:       db,
+		closeMut: &sync.RWMutex{},
+		iterWG:   sync.WaitGroup{},
+	})
+}
+
+func (db *goleveldb) Get(key []byte) ([]byte, error) {
+	return db.DB.Get(key, nil)
+}
+
+func (db *goleveldb) Has(key []byte) (bool, error) {
+	return db.DB.Has(key, nil)
+}
+
+func (db *goleveldb) Put(key, val []byte) error {
+	db.closeMut.RLock()
+	defer db.closeMut.RUnlock()
+	if db.closed {
+		return leveldb.ErrClosed
+	}
+	return db.DB.Put(key, val, nil)
+}
+
+func (db *goleveldb) Delete(key []byte) error {
+	db.closeMut.RLock()
+	defer db.closeMut.RUnlock()
+	if db.closed {
+		return leveldb.ErrClosed
+	}
+	return db.DB.Delete(key, nil)
+}
+
+func (db *goleveldb) IsClosed(err error) bool {
+	return err == leveldb.ErrClosed
+}
```

Comment:
I dislike this. net package returns net.Error that has IsTimeout() etc, we should have db.Error,

or db.ErrNotFound (oppose to some checker magic).